src/hotspot/share/code/codeCache.cpp

```diff
@@ -310,9 +310,22 @@ void CodeCache::initialize_heaps() {
   FLAG_SET_ERGO(ProfiledCodeHeapSize, profiled_size);
   FLAG_SET_ERGO(NonProfiledCodeHeapSize, non_profiled_size);
 
+  const size_t ps = page_size(false, 8);
+  // Print warning if using large pages but not able to use the size given
+  if (UseLargePages && ps < LargePageSizeInBytes) {
+      char msg[256];
+      jio_snprintf(msg, sizeof(msg),
+                   "Failed to reserve large page memory for code cache (" SIZE_FORMAT "%s). "
+                   "Reverting to smaller page size (" SIZE_FORMAT "%s).",
+                   byte_size_in_exact_unit(LargePageSizeInBytes), exact_unit_for_byte_size(LargePageSizeInBytes),
+                   byte_size_in_exact_unit(ps), exact_unit_for_byte_size(ps));
+      log_warning(codecache)("%s", msg);
+      warning("%s", msg);
+  }
```

Comment:
This "warn if page size is not as expected" topic is more complex.

For a start, the text is misleading. You don't reserve anything here. "reserve" has a very clear meaning. Here, you only calculated a page size that fits CodeCacheSize geometry, and then found it is smaller than what the user probably wanted, and you want to warn the user about this. Makes sense.

We also have the following cases:

- code cache could, in theory, be satisfied with large pages, but its size is not aligned with large page size. E.g. 2MB pages and CodeCacheSize=101m, would result in code cache using 4KB pages.
- The user may not have LPSIB specified, so it's 0, but he specified +UseLargePages. Then, he may not have a specific page size in mind but may still want to know if the page size of the code cache is not a large page. 

I think if you warn here when we divert from planned page size for geometry reasons, you should warn for these cases too. An acceptable minimum would be "ps < os::default_large_page_size()"

Reply:
Right, thanks!

There is just this point that it is not completely clear to me:
> * code cache could, in theory, be satisfied with large pages, but its size is not aligned with large page size. E.g. 2MB pages and CodeCacheSize=101m, would result in code cache using 4KB pages.

Do you mean that with such a configuration (2MB pages and CodeCacheSize=101m) would the code cache use 4KB pages anyway later on? Wouldn't `CodeCache::page_size` return 2MB pages (and possibly align the code cache to 102MB)? (I guess I'm missing something here)

Reply:
I may be wrong too. Looking closer, I think it aligns up. So, I think you can forget the first of the two points.

Reply:
I've changed "reserve" into the more generic "use" in the message.
To check for the actual large page size I've used the same `page_size` method giving 1 as minimum number of pages (`default_large_page_size` seems to be specific for linux)

Reply:
> Sorry to be a stickler for precise warning messages. Proposal: "Code cache size too small for " PROPER_FMT " pages." - the rest is fine.

Oh, I see! I saw the issue the other way around 🙂. The message has been changed. Thanks!